### PR TITLE
feat: plan part for new s3 source

### PR DESCRIPTION
**Pull request by tabVersion**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

part of #12410

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/logical_source.rs [hunk 1]
```diff
@@ -152,6 +159,50 @@ impl LogicalSource {
...
+                    .as_ref()
+                    .map_or(true, |s| s.append_only),
+                false,
+                FixedBitSet::with_capacity(logical_source.column_catalog.len()),
+            ),
+            logical: logical_source,
+        }
+        .into();
+        new_s3_plan = RequiredDist::shard_by_key(1, &[0])
+            .enforce_if_not_satisfies(new_s3_plan, &Order::any())?;
```
Comment (tabVersion) [inline src/frontend/src/optimizer/plan_node/logical_source.rs:196]: @st1page any wrong with this part? I got a panic when explaining the query `explain create table t (v1 int, v2 varchar ) with ( connector = 's3_v2') format plain encode json ;`

Comment (st1page) [reply src/frontend/src/optimizer/plan_node/logical_source.rs:196]: I will take a look

**Comment by tabVersion** [comment]
image
It works as expected now

File: src/connector/src/source/base.rs [hunk 2]
```diff
@@ -296,9 +298,17 @@ pub trait SplitReader: Sized + Send {
 
 for_all_sources!(impl_connector_properties);
 
+impl ConnectorProperties {
+    pub fn is_new_fs_connector(props: &BTreeMap<String, String>) -> bool {
+        props
+            .get(UPSTREAM_SOURCE_KEY)
+            .map(|s| s.eq_ignore_ascii_case(S3_V2_CONNECTOR))
```
Comment (st1page) [inline src/connector/src/source/base.rs:305]: a little out of topic, will we support other file system in future? can we reuse the code easily?

File: src/frontend/src/optimizer/plan_node/logical_source.rs [hunk 3]
```diff
@@ -453,18 +508,34 @@ impl PredicatePushdown for LogicalSource {
 
 impl ToBatch for LogicalSource {
     fn to_batch(&self) -> Result<PlanRef> {
+        if self.core.catalog.is_some()
+            && ConnectorProperties::is_new_fs_connector(
+                &self.core.catalog.as_ref().unwrap().properties,
+            )
+        {
+            todo!()
+        }
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/logical_source.rs:517]: prefer return an error here.

Comment (tabVersion) [reply src/frontend/src/optimizer/plan_node/logical_source.rs]: fixed

File: src/frontend/src/optimizer/plan_node/logical_source.rs [hunk 4]
```diff
@@ -453,18 +508,37 @@ impl PredicatePushdown for LogicalSource {
...
         let source = self.wrap_with_optional_generated_columns_batch_proj()?;
         Ok(source)
     }
 }
 
 impl ToStream for LogicalSource {
     fn to_stream(&self, _ctx: &mut ToStreamContext) -> Result<PlanRef> {
-        let mut plan = if self.core.for_table {
-            StreamSource::new(self.rewrite_to_stream_batch_source()).into()
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/logical_source.rs:464]: this logic is lost

Comment (tabVersion) [reply src/frontend/src/optimizer/plan_node/logical_source.rs:464]: fixed

**Comment by tabVersion** [comment]
contained in #12595

**Pull request closed without merging**